### PR TITLE
Do not re-sequence replayed transactions

### PR DESCRIPTION
**Pull request by asonnino**
The authority currently sequences every shared-object transaction it sees. This is fine, but we can skip consensus in the following cases:
* we already executed the transaction, we can thus immeditaly return the effects without first going through consensus
* we already assigned locks to the transaction but failed to execute it. This scenario happens when the authority missed some of the transaction's dependencies; we can thus try to re-execute it now.

### CODE REVIEW

**Review by lxfind — APPROVED** [review]

File: sui_core/src/authority_server.rs [hunk 1]
```diff
@@ -212,12 +212,24 @@ impl Validator for AuthorityServer {
...
-            .consensus_adapter
-            .submit(&transaction)
+        // In some cases we can skip consensus for shared-object transactions: (i) we already executed
+        // the transaction, (ii) we already assigned locks to the transaction but failed to execute it.
+        // The later scenario happens when the authority missed some of the transaction's dependencies;
+        // we can thus try to re-execute it now.
+        let ConsensusTransaction::UserTransaction(certificate) = &transaction;
+        let info = match self
+            .state
+            .try_skip_consensus(certificate)
```
Comment (huitseeker) [inline sui_core/src/authority_server.rs:222]: You own the transaction, consider having `try_skip_consensus` take ownership of it.
See [C-CALLER-CONTROL](https://rust-lang.github.io/api-guidelines/flexibility.html#caller-decides-where-to-copy-and-place-data-c-caller-control).

**Review by huitseeker — APPROVED** [review]
A few nits, but this looks great!

File: sui_types/src/error.rs [hunk 2]
```diff
@@ -318,8 +318,10 @@ pub enum SuiError {
     SharedObjectLockingFailure(String),
     #[error("Consensus listener is out of capacity")]
     ListenerCapacityExceeded,
-    #[error("Failed to  serialize/deserialize Narwhal message: {0}")]
+    #[error("Failed to serialize/deserialize Narwhal message: {0}")]
     ConsensusNarwhalSerializationError(String),
+    #[error("Only shared object transactions need to be sequenced")]
+    OnlySharedObjectTransactionNeedSequencing,
```
[the remark this answers is not in the document]
Comment (asonnino) [reply sui_types/src/error.rs:324]: good idea

File: sui_core/src/authority_server.rs (hunk 1 above)
Comment (asonnino) [reply sui_core/src/authority_server.rs:222]: thanks!

Comment (asonnino) [reply sui_core/src/authority_server.rs:222]: Actually this is a good read (should have done it earlier): https://rust-lang.github.io/api-guidelines/about.html

**Pull request merged**